homeassistant/components/ring/coordinator.py

```diff
@@ -0,0 +1,105 @@
+"""Data coordinators for the ring integration."""
+from collections.abc import Callable
+from dataclasses import dataclass
+import logging
+from typing import Any, Optional
+
+import ring_doorbell
+from ring_doorbell.generic import RingGeneric
+
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import NOTIFICATIONS_SCAN_INTERVAL, SCAN_INTERVAL
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def _call_api(
+    hass: HomeAssistant, target: Callable[..., Any], *args, msg_suffix: str = ""
+):
+    try:
+        return await hass.async_add_executor_job(target, *args)
+    except ring_doorbell.AuthenticationError as err:
+        # Raising ConfigEntryAuthFailed will cancel future updates
+        # and start a config flow with SOURCE_REAUTH (async_step_reauth)
+        raise ConfigEntryAuthFailed from err
+    except ring_doorbell.RingTimeout as err:
+        raise UpdateFailed(
+            f"Timeout communicating with API{msg_suffix}: {err}"
+        ) from err
+    except ring_doorbell.RingError as err:
+        raise UpdateFailed(f"Error communicating with API{msg_suffix}: {err}") from err
+
+
+@dataclass
+class RingDeviceData:
+    """RingDeviceData."""
+
+    device: RingGeneric
+    history: Optional[list] = None
+
+
+class RingDataCoordinator(DataUpdateCoordinator[dict[int, RingDeviceData]]):
+    """Base class for device coordinators."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        ring_api: ring_doorbell.Ring,
+    ) -> None:
+        """Initialize my coordinator."""
+        super().__init__(
+            hass,
+            name="devices",
+            logger=_LOGGER,
+            update_interval=SCAN_INTERVAL,
+        )
+        self.ring_api: ring_doorbell.Ring = ring_api
+        self.first_call: bool = True
+
+    async def _async_update_data(self):
+        """Fetch data from API endpoint."""
+        update_method: str = "update_data" if self.first_call else "update_devices"
+        await _call_api(self.hass, getattr(self.ring_api, update_method))
+        self.first_call = False
+        data: dict[str, RingDeviceData] = {}
+        devices: dict[str : list[RingGeneric]] = self.ring_api.devices()
+        for device_type in devices:
+            for device in devices[device_type]:
```

Comment:
```suggestion
            subscribed_device_ids = set(self.async_contexts())
            for device in devices[device_type]:
```
So we do it only once instead for each device

Reply:
Done